File: src/main/java/nz/co/searchwellington/utils/RobotsAwareHttpFetcher.java
package nz.co.searchwellington.utils;

import nz.co.searchwellington.http.HttpFetcher;
import org.apache.commons.httpclient.HttpStatus;
import org.apache.log4j.Logger;

public class RobotsAwareHttpFetcher implements HttpFetcher {
	
    private Logger log = Logger.getLogger(RobotsAwareHttpFetcher.class);

	private RobotExclusionService robotExclusionService;
	private HttpFetcher httpFetcher;
	private String[] exceptions;
		
	public RobotsAwareHttpFetcher(RobotExclusionService robotExclusionService, HttpFetcher httpFetcher, String... exceptions) {
		this.robotExclusionService = robotExclusionService;
		this.httpFetcher = httpFetcher;
		this.exceptions = exceptions;
	}

	public HttpFetchResult httpFetch(String url) {		
		boolean overrideRobotDotTxt = false;
		for (String exception : exceptions) {
			if (url.startsWith(exception)) {
				overrideRobotDotTxt = true;
			}
		}
		
		if (overrideRobotDotTxt || robotExclusionService.isUrlCrawlable(url, getUserAgent())) {
			return httpFetcher.httpFetch(url);
		}
		log.info("Url is not allowed to be crawled: " + url);
		return new HttpFetchResult(HttpStatus.SC_UNAUTHORIZED, null);
	}

	@Override
	public String getUserAgent() {
		return httpFetcher.getUserAgent();
	}

}

File: src/main/scala/nz/co/searchwellington/http/RobotsAwareHttpFetcher.scala
package nz.co.searchwellington.http

import java.net.URL

import org.apache.http.HttpStatus
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.{ExecutionContext, Future}

@Component
class RobotsAwareHttpFetcher @Autowired()(robotExclusionService: RobotExclusionService, httpFetcher: WSHttpFetcher) extends HttpFetcher {

  private val log = LogFactory.getLog(classOf[RobotsAwareHttpFetcher])

  private val excludedUrlPrefixes = Seq.empty

  override def httpFetch(url: URL, followRedirects: Boolean = true)(implicit ec: ExecutionContext): Future[HttpFetchResult] = {
    val ignoreRobotDotTxt = excludedUrlPrefixes.exists(e => url.toExternalForm.startsWith(e))
    if (ignoreRobotDotTxt || robotExclusionService.isUrlCrawlable(url, getUserAgent)) {
      httpFetcher.httpFetch(url, followRedirects)

    } else {
      log.info("Url is not allowed to be crawled: " + url)
      Future.successful(HttpFetchResult(HttpStatus.SC_UNAUTHORIZED, null))
    }
  }

  override def getUserAgent: String = httpFetcher.getUserAgent

}